nixos/modules/services/networking/nebula.nix

```diff
@@ -0,0 +1,187 @@
+    };
+
+    listen.host = mkOption {
+      type = types.str;
+      default = "0.0.0.0";
+      description = "IP address to listen on.";
+    };
+
+    listen.port = mkOption {
+      type = types.port;
+      default = 4242;
+      description = "Port number to listen on.";
+    };
+
+    punch = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        Continues to punch inbound/outbound at a regular interval to avoid expiration of firewall nat mappings.
+      '';
+    };
+
+    tun.disable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        When tun is disabled, a lighthouse can be started without a local tun interface (and therefore without root).
+      '';
+    };
+
+    tun.device = mkOption {
+      type = types.str;
+      default = "nebula1";
+      description = "Name of the tun device.";
+    };
+
+    firewall.outbound = mkOption {
+      type = types.listOf types.attrs;
+      default = [];
+      description = "Firewall rules for outbound traffic.";
+      example = ''[ { port = "any"; proto = "any"; host = "any"; } ]'';
+    };
+
+    firewall.inbound = mkOption {
+      type = types.listOf types.attrs;
+      default = [];
+      description = "Firewall rules for inbound traffic.";
+      example = ''[ { port = "any"; proto = "any"; host = "any"; } ]'';
+    };
+
+    extraConfig = mkOption {
+      type = types.attrs;
+      default = {};
+      description = "Extra configuration added to the Nebula config file.";
+      example = literalExample ''
+        {
+          lighthouse.dns = {
+            host = "0.0.0.0";
+            port = 53;
+          };
+        }
+      '';
+    };
+  };
+
+  # Implementation
+
+  config =
+    let
+      # Convert the options to an attrset.
+      optionConfigSet = {
+        pki = { ca = cfg.ca; cert = cfg.cert; key = cfg.key; };
+        static_host_map = cfg.staticHostMap;
+        lighthouse = { am_lighthouse = cfg.isLighthouse; hosts = cfg.lighthouses; };
+        listen = { host = cfg.listen.host; port = cfg.listen.port; };
+        punchy = { punch = cfg.punch; };
+        tun = { disabled = cfg.tun.disable; dev = cfg.tun.device; };
+        firewall = { inbound = cfg.firewall.inbound; outbound = cfg.firewall.outbound; };
+      };
+
+      # Merge in the extraconfig attrset.
+      mergedConfigSet = lib.recursiveUpdate optionConfigSet cfg.extraConfig;
+
+      # Dump the config to JSON. Because Nebula reads YAML and YAML is a superset of JSON, this works.
+      nebulaConfig = pkgs.writeTextFile { name = "nebula-config.yml"; text = (builtins.toJSON mergedConfigSet); };
+
+      # The service needs to launch as root to access the tun device, if it's enabled.
+      serviceUser = if cfg.tun.disable then "nebula" else "root";
+      serviceGroup = if cfg.tun.disable then "nebula" else "root";
+
+    in mkIf cfg.enable {
+      # Create systemd service for Nebula.
+      systemd.services.nebula = {
```

Comment:
There seems to be some disagreement between your `systemd` unit and the [ones](https://github.com/slackhq/nebula/blob/master/examples/service_scripts/nebula.service) [upstream](https://github.com/slackhq/nebula/blob/master/examples/quickstart-vagrant/ansible/roles/nebula/files/systemd.nebula.service) [provides](https://github.com/slackhq/nebula/blob/master/dist/arch/nebula.service).

Reply:
Yes, we did not use the upstream-provided units as a basis. We probably should. Let me fix that.

Reply:
I updated the unit a bit to match [this unit](https://github.com/slackhq/nebula/blob/master/examples/service_scripts/nebula.service) more closely.